### PR TITLE
Serialize model config

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

File: include/deepworks/layer_info.hpp [hunk 1]
```diff
@@ -11,18 +11,63 @@
...
+          type (AttrTraits<typename std::decay<T>::type>::type),
+          shape(AttrTraits<typename std::decay<T>::type>::shape) {
+    }
 
     template <typename T>
     const T& get() const {
         return *std::any_cast<T>(&m_value);
     }
 
+    AttrType  type;
```
Comment (TolyaTalamanov) [inline include/deepworks/layer_info.hpp:68]: Move to private and provide getters

File: include/deepworks/layer_info.hpp [hunk 2]
```diff
@@ -11,18 +11,63 @@
...
+    AttrShape shape;
```
Comment (TolyaTalamanov) [inline include/deepworks/layer_info.hpp:69]: It can be renamed possibly

File: src/expression/nn.cpp [hunk 3]
```diff
@@ -1,11 +1,13 @@
 #include <numeric>
+#include <functional>
 
 #include <deepworks/nn.hpp>
 #include <deepworks/call.hpp>
 #include <deepworks/tensor.hpp>
 #include <deepworks/initializers.hpp>
 
 #include "util/assert.hpp"
+#include <iostream>
```
Comment (TolyaTalamanov) [inline src/expression/nn.cpp:10]: remove

File: src/expression/nn.cpp [hunk 4]
```diff
@@ -17,6 +19,79 @@ dw::Linear::Linear(int units, std::string name)
     m_info.impl().attrs["units"] = units;
 }
 
+template<class T>
+struct get_in {
+    static T get(const std::vector<const dw::Attribute*>& args, int idx) {
```
Comment (TolyaTalamanov) [inline src/expression/nn.cpp:24]: I believe there is no sense to create vector, possible to get attributes directly from map based on order vector

File: src/expression/nn.cpp [hunk 5]
```diff
@@ -17,6 +19,79 @@ dw::Linear::Linear(int units, std::string name)
...
+
+    auto make_softmax = [&]{
+        return make_typed_layer<dw::Softmax>()(attrs, name)(inputs);
+    };
+
+    table_t supported_layers = {
+        {"Linear"     , make_linear},
+        {"ReLU"       , make_relu},
+        {"BatchNorm1D", make_batchnorm1d},
+        {"Softmax"    , make_softmax}
```
Comment (TolyaTalamanov) [inline src/expression/nn.cpp:84]: Add other layers

**Comment by TolyaTalamanov** [comment]
@NeuralDeepWorks/maintainers Have a look, please

File: include/deepworks/base_op.hpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <deepworks/placeholder.hpp>
+#include <deepworks/call.hpp>
+#include <deepworks/layer_info.hpp>
+
+namespace deepworks {
+
+template <typename R, class C, typename ... Types>
+constexpr size_t get_num_ins( R(C::*)(Types ...)) {
```
Comment (apertovs) [inline include/deepworks/base_op.hpp:10]: I think get_number_of_arguments is more clear

File: include/deepworks/model.hpp [hunk 7]
```diff
@@ -47,4 +52,32 @@ class Model {
...
+        std::string name;
+        std::string type;
+        Attributes  attrs;
+
+        std::vector<int> in_ids;
+        std::vector<int> out_ids;
+    };
+
+    std::vector<int> input_ids;
+    std::vector<int> output_ids;;
```
Comment (apertovs) [inline include/deepworks/model.hpp:78]: double ;

File: src/model/model.cpp [hunk 8]
```diff
@@ -11,42 +14,90 @@
...
+
+namespace dw = deepworks;
+using namespace dw::graph;
+
+struct is_data {
+   bool operator()(ade::NodeHandle nh) {
+       return tgraph.metadata(nh).get<Type>().t == Type::DATA;
+   }
+
+   TypedGraph& tgraph;
```
Comment (apertovs) [inline src/model/model.cpp:28]: Can we move this reference to operator arguments?

File: src/model/model.cpp [hunk 9]
```diff
@@ -11,42 +14,90 @@
...
+};
+
+struct is_op {
+   bool operator()(ade::NodeHandle nh) {
+       return tgraph.metadata(nh).get<Type>().t == Type::OP;
+   }
+
+   TypedGraph& tgraph;
```
Comment (apertovs) [inline src/model/model.cpp:36]: Can we move this reference to operator arguments?

**Review by apertovs — APPROVED** [review]
Good job!

File: include/deepworks/base_op.hpp (hunk 6 above)
Comment (TolyaTalamanov) [reply include/deepworks/base_op.hpp:10]: and it's much longer as well :)

File: src/model/model.cpp (hunk 8 above)
Comment (TolyaTalamanov) [reply src/model/model.cpp:28]: It's done in order to reduce number of codes

File: include/deepworks/base_op.hpp (hunk 6 above)
Comment (TolyaTalamanov) [reply include/deepworks/base_op.hpp:10]: get_num_args is Ok ?

File: src/model/model.cpp (hunk 9 above)
Comment (TolyaTalamanov) [reply src/model/model.cpp:36]: t's done in order to reduce number of codes

File: src/expression/nn.cpp (hunk 5 above)
Comment (TolyaTalamanov) [reply src/expression/nn.cpp]: done

File: src/expression/nn.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply src/expression/nn.cpp]: done

File: include/deepworks/model.hpp (hunk 7 above)
Comment (TolyaTalamanov) [reply include/deepworks/model.hpp]: Done

File: include/deepworks/base_op.hpp (hunk 6 above)
Comment (TolyaTalamanov) [reply include/deepworks/base_op.hpp]: Done

File: src/expression/nn.cpp (hunk 3 above)
Comment (TolyaTalamanov) [reply src/expression/nn.cpp]: Done

File: include/deepworks/layer_info.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply include/deepworks/layer_info.hpp]: Done

File: src/expression/nn.cpp [hunk 10]
```diff
@@ -17,6 +18,112 @@ dw::Linear::Linear(int units, std::string name)
...
+        return make_typed_layer<dw::Dropout, float>({"p"})(attrs, name)(inputs);
+    };
+
+    table_t supported_layers = {
+        {"Linear"     , make_linear},
+        {"ReLU"       , make_relu},
+        {"BatchNorm1D", make_batchnorm1d},
+        {"Softmax"    , make_softmax},
+        {"MaxPooling" , make_maxpooling},
+        {"MaxPooling" , make_convolution},
```
Comment (l-bat) [inline src/expression/nn.cpp:112]:
```suggestion
        {"Convolution" , make_convolution},
```

File: src/model/model.cpp [hunk 11]
```diff
@@ -156,28 +208,59 @@ void deepworks::Model::compile() {
 }
 
 void deepworks::Model::forward (const deepworks::Tensor& input,
-                                      deepworks::Tensor& output) {
+        deepworks::Tensor& output) {
```
Comment (l-bat) [inline src/model/model.cpp:211]: Fix indentation

Comment (TolyaTalamanov) [reply src/model/model.cpp]: Done

**Review by l-bat — APPROVED** [review]
Thank you!

**Pull request merged**